llvm/lib/Target/AIE/AIEWawRegRewriter.cpp

```diff
@@ -0,0 +1,347 @@
+
+  bool isLastDef(const MachineOperand &Def, const MachineInstr &DefInstr,
+                 const MachineBasicBlock &MBB);
+
+  void addAliasRegs(BitVector &ForbiddenPhysRegs, const MCPhysReg &PhysReg);
+};
+
+MCPhysReg
+AIEWawRegRewriter::getAssignedPhysReg(const MachineOperand &MO) const {
+  Register Reg = MO.getReg();
+
+  if (Reg.isStack())
+    return MCRegister::NoRegister;
+
+  if (Reg.isVirtual())
+    Reg = VRM->getPhys(Reg);
+
+  return Reg;
+}
+
+bool AIEWawRegRewriter::runOnMachineFunction(MachineFunction &MF) {
+
+  MachineLoopInfo &MLI = getAnalysis<MachineLoopInfo>();
+
+  SmallVector<const MachineBasicBlock *, 4> LoopMBBs =
+      AIELoopUtils::getInnermostLoopBlocks(MLI);
+
+  if (LoopMBBs.empty())
+    return false;
+
+  this->MF = &MF;
+  MRI = &MF.getRegInfo();
+  TRI = static_cast<const AIEBaseRegisterInfo *>(MRI->getTargetRegisterInfo());
+  VRM = &getAnalysis<VirtRegMap>();
+  LRM = &getAnalysis<LiveRegMatrix>();
+  LIS = &getAnalysis<LiveIntervals>();
+  Modified = false;
+
+  LLVM_DEBUG(dbgs() << "*** WAW Loop Register Rewriting: " << MF.getName());
+  LLVM_DEBUG(dbgs() << " ***\n");
+
+  for (const MachineBasicBlock *MBB : LoopMBBs)
+    renameMBBPhysRegs(MBB);
+
+  return Modified;
+}
+
+// return a vector containing virtual registers, that have had a copy to define
+// the virtual register
+SmallVector<Register, 16> getVRegWithCopies(const MachineBasicBlock &MBB) {
+  SmallVector<Register, 16> VRegWithCopies;
+  for (const MachineInstr &MI : MBB) {
+    for (const MachineOperand Def : MI.defs()) {
+      Register Reg = Def.getReg();
+
+      if (!Reg.isVirtual())
+        continue;
+
+      if (MI.isCopyLike() && find(VRegWithCopies, Reg) == VRegWithCopies.end())
+        VRegWithCopies.push_back(Reg);
+    }
+  }
+
+  return VRegWithCopies;
+}
+
+bool AIEWawRegRewriter::renameMBBPhysRegs(const MachineBasicBlock *MBB) {
+  LLVM_DEBUG(dbgs() << "WAW Reg Renaming BasicBlock "; MBB->dump();
+             dbgs() << "\n");
+  BitVector UsedPhysRegs(TRI->getNumRegs());
+  SmallVector<std::pair<MCPhysReg, MCPhysReg>, 8> ReplacementRegs;
+
+  BitVector BlockedPhysRegs = getDefinedPhysRegs(MBB);
+  SmallVector<Register, 16> VRegWithCopies = getVRegWithCopies(*MBB);
+
+  for (const MachineInstr &MI : *MBB) {
+
+    // phi nodes that will be eliminated, do not define a physical register
+    // since they will be eliminated in a later pass, and the copy will be
+    // removed
+    if (isIdentityCopy(MI))
+      continue;
+
+    for (const MachineOperand &MO : MI.defs()) {
+
+      Register Reg = MO.getReg();
+      if (!Reg.isVirtual())
+        continue;
+      if (VRM->hasRequiredPhys(Reg))
+        continue;
+      if (MO.isTied())
+        continue;
+      if (!TRI->isVecOrAccRegClass(*(MRI->getRegClass(Reg))))
+        continue;
+      if (!isLastDef(MO, MI, *MBB))
+        continue;
+
+      MCPhysReg PreSwapPhysReg = VRM->getPhys(Reg);
+      if (isWorthRenaming(Reg, UsedPhysRegs, VRegWithCopies) &&
```

Comment:
Since UsedPhysregs is basic-block local, the effect it has on multi-block loops is questionable. Perhaps make it absolutely clear that it is intended for single block loops by only running it on single block loops.

Reply:
I modified `AIELoopUtils::getSingleBlockLoopMBBs` so that it only returns a single MBB which conforms with `singleMBBLoops`